controllers/portal_controller.go

```diff
@@ -0,0 +1,134 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corev1alpha1 "github.com/kubebb/core/api/v1alpha1"
+)
+
+// PortalReconciler reconciles a Portal object
+type PortalReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	Recorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *PortalReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Starting portal reconcile")
+
+	var portal = corev1alpha1.Portal{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &portal); err != nil {
+		if k8serror.IsNotFound(err) {
+			// Portal has been deleted.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	err := r.checkDuplicate(ctx, logger, &portal)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PortalReconciler) checkDuplicate(ctx context.Context, logger logr.Logger, portal *corev1alpha1.Portal) error {
+	list, err := r.getList(ctx, logger, portal)
+	if err != nil {
+		return err
+	}
+
+	var conflicts []string
+	for _, p := range list {
+		conflicts = append(conflicts, p.Namespace+"/"+p.Name)
+	}
+	if len(conflicts) > 0 {
+		logger.Error(errors.New("duplicate error"), "found portals with duplicate path and entry", "duplicates", conflicts)
+		portal.Status.Duplicate = true
+		portal.Status.ConflictedPortals = conflicts
```

Comment:
After setting the status of CR here, you should update the instance.

Reply:
Thank you! The instance is updated now!